// mcp_control/AccountControl.js
var async = require('async');
var dc = require('mcp_db').dc;

var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;
var dateUtil = esut.dateUtil;
var cons = require('mcp_constants');
var userType = cons.userType;
var userBillCons = cons.userBillCons;

var config = require('mcp_config');
var ec = config.ec;
var prop = config.prop;

var service = require('mcp_service');
var digestService = service.digestService;

var AccountControl = function () {
};

AccountControl.prototype.handle = function (headNode, bodyStr, userCb) {
    var self = this;
    async.waterfall([
        //是否是支持的cmd
        function (cb) {
            if (typeof self['handle' + headNode.cmd] === 'function') {
                cb(null);
            } else {
                cb(ec.E2010);
            }
        },
        //校验头的用户类型是否合法
        function (cb) {
            var userTypeId = userType[headNode.userType];
            if (userTypeId == undefined) {
                cb(ec.E9030);
            }
            else {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function (userTypeId, cb) {
            digestService.getKey({fromType: prop.digestFromType.DB, userId: headNode.userId, userType: userTypeId},
                function (err, key) {
                    cb(err, key);
                });
        },
        //先解密
        function (key, cb) {
            var decodedBodyStr = digestUtil.check(headNode, key, bodyStr);
            if (decodedBodyStr == null) {
                cb(ec.E9020);
                return;
            }
            var bodyNode;
            try {
                bodyNode = JSON.parse(decodedBodyStr);
                headNode.key = key;
            }
            catch (err) {
                cb(ec.E0005);
                return;
            }
            cb(null, bodyNode);
        },
        //check the param
        function (bodyNode, cb) {
            var method = 'check' + headNode.cmd;
            self[method](headNode, bodyNode, function (err) {
                cb(err, bodyNode);
            });
        },
        //业务处理
        function (bodyNode, cb) {
            var cmd = 'handle' + headNode.cmd;
            self[cmd](headNode, bodyNode, cb);
        }
    ], function (err, bodyNode) {
        userCb(err, bodyNode);
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC01 账户记录查询
 */

AccountControl.prototype.checkAC01 = function (headNode, bodyNode, cb) {
    cb(null);
};


AccountControl.prototype.handleAC01 = function (headNode, bodyNode, cb) {
    var termTable = dc.mg.get("userbill");
    var cursor = termTable.find(bodyNode.cond, bodyNode.fields).sort(bodyNode.sort).skip(bodyNode.skip).limit(bodyNode.limit);
    cursor.toArray(function (err, data) {
        for (var key in data) {
            var set = data[key];
            set.create_time = dateUtil.toString(set.create_time);
        }
        var backBodyNode = {};
        if (data != null && data != undefined) {
            backBodyNode.rst = data;
        }

        backBodyNode.count = cursor.count(function (err, count) {
            backBodyNode.count = count;
            cb(null, backBodyNode);
        });
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC02 账户操作
 */

AccountControl.prototype.checkAC02 = function (headNode, bodyNode, cb) {
    cb(null);
};


AccountControl.prototype.handleAC02 = function (headNode, bodyNode, cb) {
    var accountTab = dc.main.get('account');
    async.waterfall([
        //获取余额信息
        function (cb) {
            accountTab.findOne({user_id: bodyNode.cond.user_id}, {}, [], function (err, accountObj) {
                cb(null, accountObj);
            })
        },
        function (accountObj, cb) {
            var money = accountObj.money;
            var frozen_money = accountObj.frozen_money;
            bodyNode.cond.proMoney = money;
            switch (bodyNode.type) {
                //增加余额
                case userBillCons.transType.IN:
                    //余额
                    money = money + bodyNode.cond.money;
                    //记录
                    bodyNode.cond.afterMoney = money;
                    bodyNode.cond.frozenMoney = frozen_money;
                    cb(null, money, frozen_money, bodyNode.cond);
                    break;
                case userBillCons.transType.OUT:
                    //余额
                    money = money - bodyNode.cond.money;
                    //记录
                    bodyNode.cond.afterMoney = money;
                    bodyNode.cond.frozenMoney = frozen_money;
                    cb(null, money, frozen_money, bodyNode.cond);
                    break;
                case userBillCons.transType.FROZEN:
                    //余额
                    money = money - bodyNode.cond.money;
                    frozen_money = frozen_money + bodyNode.cond.money;
                    //记录
                    bodyNode.cond.afterMoney = money;
                    bodyNode.cond.frozenMoney = frozen_money;
                    cb(null, money, frozen_money, bodyNode.cond);
                    break;
                case userBillCons.transType.THAW:
                    //余额
                    money = money + bodyNode.cond.money;
                    frozen_money = frozen_money - bodyNode.cond.money;
                    //记录
                    bodyNode.cond.afterMoney = money;
                    bodyNode.cond.frozenMoney = frozen_money;
                    cb(null, money, frozen_money, bodyNode.cond);
                    break;
                case userBillCons.transType.FROZEN_OUT:
                    frozen_money = frozen_money - bodyNode.cond.money;
                    //记录
                    bodyNode.cond.afterMoney = money;
                    bodyNode.cond.frozenMoney = frozen_money;
                    cb(null, money, frozen_money, bodyNode.cond);
                    break;
                default:
                    cb(ec.E3070);
            }
        },
        function (money, frozen_money, userBill) {
            var userBillTab = dc.mg.get('userbill');
            userBill.create_time = new Date().getTime();
            userBillTab.save(userBill, [], function (err, _r1) {
                accountTab.update({user_id: userBill.user_id}, {
                    $set: {
                        money: money,
                        frozen_money: frozen_money,
                        update_time: dateUtil.getCurTime()
                    }
                }, [], function (err, _r2) {
                    if (err) {
                        cb(ec.E9999);
                    } else {
                        cb(null);
                    }
                })
            });
        }
    ], function (err) {
        cb(err);
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC03 查询账户信息
 */

AccountControl.prototype.checkAC03 = function (headNode, bodyNode, cb) {
    cb(null);
};


AccountControl.prototype.handleAC03 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    async.waterfall([
        //第一步获取待收项
        function (cb) {
            var noRepaySql = "SELECT sum(a.money) as allMoney,sum(a.interest) as allInterest,sum(a.st_interest) as AllStInterest" +
                " from invest_repay a " +
                " where a.invest_id in (select id from invest b where b.user_id =(SELECT id from user c where c.username='" + bodyNode.username + "')) and a.status=0";
            conn.execute(noRepaySql, [], function (err, data) {
                if (err) {
                    cb(ec.E9999);
                } else {
                    if (data.length == 1) {
                        if (data[0].allMoney == undefined || data[0].allMoney == null) {
                            backNode.entity.noAllMoney = 0;
                        } else {
                            backNode.entity.noAllMoney = data[0].allMoney;
                        }
                        if (data[0].allInterest == undefined || data[0].allInterest == null) {
                            backNode.entity.noAllInterest = 0;
                        } else {
                            backNode.entity.noAllInterest = data[0].allInterest;
                        }
                        if (data[0].AllStInterest == undefined || data[0].AllStInterest == null) {
                            backNode.entity.noAllStInterest = 0;
                        } else {
                            backNode.entity.noAllStInterest = data[0].AllStInterest;
                        }
                    } else {
                        backNode.entity.noAllMoney = 0;
                        backNode.entity.noAllInterest = 0;
                        backNode.entity.noAllStInterest = 0;
                    }
                    cb(null);
                }
            })

        },
        //第二步获取已收项
        function (cb) {
            //已收
            var hadRepaySql = "SELECT sum(a.money) as allMoney,sum(a.interest) as allInterest,sum(st_interest) as AllStInterest" +
                " from invest_repay a " +
                " where a.invest_id in (select id from invest b where b.user_id =(SELECT id from user c where c.username='" + bodyNode.username + "')) and status=1";
            conn.execute(hadRepaySql, [], function (err, data) {
                if (err) {
                    cb(ec.E9999);
                } else {
                    if (data.length == 1) {
                        console.log(data);
                        if (data[0].allMoney == undefined || data[0].allMoney == null) {
                            backNode.entity.hadAllMoney = 0;
                        } else {
                            backNode.entity.hadAllMoney = data[0].allMoney;
                        }
                        if (data[0].allInterest == undefined || data[0].allInterest == null) {
                            backNode.entity.hadAllInterest = 0;
                        } else {
                            backNode.entity.hadAllInterest = data[0].allInterest;
                        }
                        if (data[0].AllStInterest == undefined || data[0].AllStInterest == null) {
                            backNode.entity.hadAllStInterest = 0;
                        } else {
                            backNode.entity.hadAllStInterest = data[0].AllStInterest;
                        }
                    } else {
                        backNode.entity.hadAllMoney = 0;
                        backNode.entity.hadAllInterest = 0;
                        backNode.entity.hadAllStInterest = 0;
                    }
                    cb(null);
                }
            })
        },
        //第三步获取账户信息
        function (cb) {
            var accountSql = "SELECT money,frozen_money from account where user_id=(SELECT id from user where username='" + bodyNode.username + "')";
            conn.execute(accountSql, [], function (err, data) {
                if (err) {
                    cb(ec.E9999);
                } else {
                    if (data.length == 1) {
                        if (data[0].money == undefined || data[0].money == null) {
                            backNode.entity.money = 0;
                        } else {
                            backNode.entity.money = data[0].money;
                        }
                        if (data[0].frozen_money == undefined || data[0].frozen_money == null) {
                            backNode.entity.frozen_money = 0;
                        } else {
                            backNode.entity.frozen_money = data[0].frozen_money;
                        }
                    } else {
                        backNode.entity.money = 0;
                        backNode.entity.frozen_money = 0;
                    }
                    cb(null);
                }
            })
        }
    ], function (err) {
        if (err) {
            cb(err);
        } else {
            cb(null, backNode);
        }
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC04 查询账户总投资和总收益
 */

AccountControl.prototype.checkAC04 = function (headNode, bodyNode, cb) {
    if (bodyNode.userId == undefined || bodyNode.userId == '') {
        cb(ec.E9999);
    } else {
        cb(null);
    }
};


AccountControl.prototype.handleAC04 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var typeStr = '';
    if (bodyNode.loanType != undefined && bodyNode.loanType != '') {
        typeStr = " and b.loan_type='" + bodyNode.loanType + "'";
    }
    var sql = "select sum(a.money) as money,sum(a.interest+a.st_interest) as interest from invest a LEFT JOIN loan b on a.loan_id=b.id  " +
        " where a.status=1200 and a.user_id='" + bodyNode.userId + "' " + typeStr;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
                if (data[0].interest == undefined || data[0].interest == null) {
                    backNode.entity.interest = 0;
                } else {
                    backNode.entity.interest = data[0].interest;
                }
            } else {
                backNode.entity.money = 0;
                backNode.entity.interest = 0;
            }
            cb(null, backNode);
        }
    })
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC05 查询待收本金和收益
 */

AccountControl.prototype.checkAC05 = function (headNode, bodyNode, cb) {
    if (bodyNode.userId == undefined || bodyNode.userId == '') {
        cb(ec.E9999);
    } else {
        cb(null);
    }
};

AccountControl.prototype.handleAC05 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var typeStr = '';
    if (bodyNode.loanType != undefined && bodyNode.loanType != '') {
        typeStr = " and lo.loan_type='" + bodyNode.loanType + "'";
    }
    var sql = "select sum(ire.money) as money,sum(ire.interest+ire.st_interest) as interest from invest_repay ire " +
        " LEFT JOIN invest inv on ire.invest_id=inv.id LEFT JOIN loan lo on inv.loan_id=lo.id where ire.status=0 " +
        " and inv.user_id='" + bodyNode.userId + "'"
        + typeStr;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
                if (data[0].interest == undefined || data[0].interest == null) {
                    backNode.entity.interest = 0;
                } else {
                    backNode.entity.interest = data[0].interest;
                }
            } else {
                backNode.entity.money = 0;
                backNode.entity.interest = 0;
            }
            cb(null, backNode);
        }
    })
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC06 查询待还款本金和利息
 */

AccountControl.prototype.checkAC06 = function (headNode, bodyNode, cb) {
    if (bodyNode.userId == undefined || bodyNode.userId == '') {
        cb(ec.E9999);
    } else {
        cb(null);
    }
};

AccountControl.prototype.handleAC06 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var typeStr = '';
    if (bodyNode.loanType != undefined && bodyNode.loanType != '') {
        typeStr = " and lo.loan_type='" + bodyNode.loanType + "'";
    }
    var sql = "select sum(lr.money) as money,sum(lr.interest+lr.st_interest) as interest from loan_repay lr LEFT JOIN loan lo on lr.loan_id=lo.id where lr.status=0 " +
        " and lo.repay_user_id='" + bodyNode.userId + "'"
        + typeStr;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
                if (data[0].interest == undefined || data[0].interest == null) {
                    backNode.entity.interest = 0;
                } else {
                    backNode.entity.interest = data[0].interest;
                }
            } else {
                backNode.entity.money = 0;
                backNode.entity.interest = 0;
            }
            cb(null, backNode);
        }
    })

};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC07 查询账户总投资和总收益
 */

AccountControl.prototype.checkAC07 = function (headNode, bodyNode, cb) {
    if (bodyNode.userId == undefined || bodyNode.userId == '') {
        cb(ec.E9999);
    } else {
        cb(null);
    }
};


AccountControl.prototype.handleAC07 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var typeStr = '';
    if (bodyNode.loanType != undefined && bodyNode.loanType != '') {
        typeStr = " and b.loan_type='" + bodyNode.loanType + "'";
    }
    var sql = "select sum(a.money) as money from invest a LEFT JOIN loan b on a.loan_id=b.id  " +
        " where a.status=1200 and a.user_id='" + bodyNode.userId + "' " + typeStr;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
            } else {
                backNode.entity.money = 0;
            }
            cb(null, backNode);
        }
    })
};



/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC08 查询账户投资数量
 */

AccountControl.prototype.checkAC08 = function (headNode, bodyNode, cb) {
    if (bodyNode.userId == undefined || bodyNode.userId == '') {
        cb(ec.E9999);
    } else {
        cb(null);
    }
};


AccountControl.prototype.handleAC08 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var typeStr = '';
    if (bodyNode.loanType != undefined && bodyNode.loanType != '') {
        typeStr = " and b.loan_type='" + bodyNode.loanType + "'";
    }
    var sql = "select count(*) as countNum from invest a LEFT JOIN loan b on a.loan_id=b.id  " +
        " where a.status='1200' and  a.user_id='" + bodyNode.userId + "' " + typeStr;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].countNum == undefined || data[0].countNum == null) {
                    backNode.entity.countNum = 0;
                } else {
                    backNode.entity.countNum = data[0].countNum;
                }
            } else {
                backNode.entity.countNum = 0;
            }
            cb(null, backNode);
        }
    })
};



/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC09 查询投资计划中金额
 */
AccountControl.prototype.checkAC09 = function (headNode, bodyNode, cb) {
    cb(null);
};

AccountControl.prototype.handleAC09 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var sql = "select sum(inre.money) as money,sum(inre.interest) as interest from invest_repay inre LEFT JOIN invest inv on inre.invest_id=inv.id LEFT JOIN loan lo on inv.loan_id=lo.id where 1 = 1 ";
    var condition='';
    if (bodyNode.userId != null && bodyNode.userId != '') {
        condition += " and inv.user_id='" + bodyNode.userId+"'";
    }
    if (bodyNode.status != null && bodyNode.status != '') {
        condition += " and inre.status=" + bodyNode.status;
    }
    if (bodyNode.loanId != null && bodyNode.loanId != '') {
        condition += " and lo.id = '" + bodyNode.loanId + "'";
    }
    sql+=condition;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
                if (data[0].interest == undefined || data[0].interest == null) {
                    backNode.entity.interest = 0;
                } else {
                    backNode.entity.interest = data[0].interest;
                }
            } else {
                backNode.entity.money = 0;
                backNode.entity.interest = 0;
            }
            cb(null, backNode);
        }
    });
};








/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  AC10 查询投资计划中待回款的本金，利息，罚息
 */
AccountControl.prototype.checkAC10 = function (headNode, bodyNode, cb) {
    cb(null);
};

AccountControl.prototype.handleAC10 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var backNode = {};
    var entity = {};
    backNode.entity = entity;
    var sql = "select  count(a.id) as nums,sum(a.money) as money,sum(a.interest+a.st_interest) as interest,sum(a.fee+a.penalty+a.advance_penalty) as penalty  " +
        "from invest_repay a LEFT JOIN invest b on a.invest_id=b.id where a.status=0 ";
    var condition=" and b.user_id='"+bodyNode.userId+"'";
    sql+=condition;
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(ec.E9999);
        } else {
            if (data.length == 1) {
                if (data[0].money == undefined || data[0].money == null) {
                    backNode.entity.money = 0;
                } else {
                    backNode.entity.money = data[0].money;
                }
                if (data[0].interest == undefined || data[0].interest == null) {
                    backNode.entity.interest = 0;
                } else {
                    backNode.entity.interest = data[0].interest;
                }
                if (data[0].penalty == undefined || data[0].penalty == null) {
                    backNode.entity.penalty = 0;
                } else {
                    backNode.entity.penalty = data[0].penalty;
                }
                if (data[0].nums == undefined || data[0].nums == null) {
                    backNode.entity.nums = 0;
                } else {
                    backNode.entity.nums = data[0].nums;
                }
            } else {
                backNode.entity.money = 0;
                backNode.entity.interest = 0;
                backNode.entity.penalty = 0;
                backNode.entity.nums = 0;
            }
            cb(null, backNode);
        }
    });
};






var accountControl = new AccountControl();
module.exports = accountControl;
