// mcp_control/InvestControl.js
var async = require('async');
var dc = require('mcp_db').dc;
var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;
var dateUtil=esut.dateUtil;

var cons = require('mcp_constants');
var userType = cons.userType;
var investStatus = cons.investStatus;
var loanStatus = cons.loanStatus;

var config = require('mcp_config');
var ec = config.ec;
var prop = config.prop;

var source = require('mcp_source');
var mqSource = source.mqSource;

var service = require('mcp_service');
var digestService = service.digestService;
var kvService = service.kvService;

var InvestControl = function () {
};

InvestControl.prototype.handle = function (headNode, bodyStr, userCb) {
    var self = this;
    async.waterfall([
        //是否是支持的cmd
        function (cb) {
            if (typeof self['handle' + headNode.cmd] === 'function') {
                cb(null);
            } else {
                cb(ec.E2010);
            }
        },
        //校验头的用户类型是否合法
        function (cb) {
            var userTypeId = userType[headNode.userType];
            if (userTypeId == undefined) {
                cb(ec.E9030);
            }
            else {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function (userTypeId, cb) {
            digestService.getKey({fromType: prop.digestFromType.DB, userId: headNode.userId, userType: userTypeId},
                function (err, key) {
                    cb(err, key);
                });
        },
        //先解密
        function (key, cb) {
            var decodedBodyStr = digestUtil.check(headNode, key, bodyStr);
            if (decodedBodyStr == null) {
                cb(ec.E9020);
                return;
            }
            var bodyNode;
            try {
                bodyNode = JSON.parse(decodedBodyStr);
                headNode.key = key;
            }
            catch (err) {
                cb(ec.E0005);
                return;
            }
            cb(null, bodyNode);
        },
        //check the param
        function (bodyNode, cb) {
            var method = 'check' + headNode.cmd;
            self[method](headNode, bodyNode, function (err) {
                cb(err, bodyNode);
            });
        },
        //业务处理
        function (bodyNode, cb) {
            var cmd = 'handle' + headNode.cmd;
            self[cmd](headNode, bodyNode, cb);
        }
    ], function (err, bodyNode) {
        userCb(err, bodyNode);
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  IC01 投资
 */

InvestControl.prototype.checkIC01 = function (headNode, bodyNode, cb) {
    var loanTab = dc.main.get('loan');
    loanTab.findOne({id: bodyNode.loan_id}, {
        status: 1,
        money: 1,
        loan_money: 1,
        rate: 1,
        st_rate: 1
    }, [], function (err, data) {
        if (data.status != loanStatus.raising) {
            cb(ec.E4010);
            return;
        }
        if (data.money - data.loan_money < bodyNode.money) {
            cb(ec.E4020);
            return;
        }
        bodyNode.rate = data.rate;
        if (data.st_rate > 0) {
            bodyNode.st_rate = data.st_rate;
        }
        bodyNode.channel = headNode.userId;
        cb(null);
    });
};

InvestControl.prototype.handleIC01 = function (headNode, bodyNode, cb) {
    async.waterfall([
        //生成id号
        function (cb) {
            kvService.getInvestId(function (err, id) {
                if (err) {
                    log.error(headNode.messageid + '获取InvestId出错');
                    cb(ec.E9999);
                } else {
                    cb(err, id);
                }
            });
        },
        //插入数据
        function (id, cb) {
            var investTab = dc.main.get('invest');
            bodyNode.create_time = new Date();
            bodyNode.status = investStatus.init;
            bodyNode.id = id;
            var userTab = dc.main.get('user');
            userTab.findOne({username: bodyNode.username}, {}, [], function (err, userObj) {
                delete bodyNode.username;
                bodyNode.user_id = userObj.id;
                investTab.save(bodyNode, [], function (err) {
                    if (err) {
                        cb(ec.E9050);
                    } else {
                        var backNode = {};
                        backNode.id = id;
                        cb(null, backNode);
                    }
                });
            });
        }
    ], function (err, backNode) {
        cb(err, backNode);
    });
};


/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  IC02 确认投资
 */

InvestControl.prototype.checkIC02 = function (headNode, bodyNode, cb) {
    cb(null);
};

InvestControl.prototype.handleIC02 = function (headNode, bodyNode, cb) {
    mqSource.send("server", JSON.stringify(bodyNode), function (err, data) {
        if (err) {
            log.error(err);
            cb(ec.E9999);
        } else {
            if (data) {
                cb(null);
            } else {
                cb(ec.E4000);
            }
        }
    });
};


InvestControl.prototype.checkIC03 = function (headNode, bodyNode, cb) {
    cb(null);
};

InvestControl.prototype.handleIC03 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var sql = "SELECT ";
    var condition = "loan.name,invest.id,invest.create_time,invest.money," +
        "(select sum(invest_repay.money)+sum(invest_repay.interest)+sum(invest_repay.st_interest) " +
        "from  invest_repay where invest_repay.status=0 and invest.id=invest_repay.invest_id)   as sumWaitMoney  ";
    var whCondition = " from invest LEFT JOIN loan  on invest.loan_id=loan.id where invest.status=1200 and (loan.status=1400 or loan.status=1500 or loan.status=1600)  and invest.user_id='" + bodyNode.id + "'";
    var exSql = sql + condition + whCondition;
    exSql += '  order by invest.create_time desc';
    exSql += '  limit ' + bodyNode.skip;
    exSql += ',' + bodyNode.limit;
    console.log(exSql);
    conn.execute(exSql, [], function (err, data) {
        for(var key in data)
        {
            var set = data[key];
            set.create_time = dateUtil.toString(set.create_time);
        }
        var backNode = {};
        backNode.rst = data;
        var countSql=sql+" count(*)  as num "+whCondition;
        conn.execute(countSql, [], function (err, countData) {
            backNode.count = countData[0].num;
            cb(null,backNode)
        });
    });
};


InvestControl.prototype.checkIC04 = function (headNode, bodyNode, cb) {
    cb(null);
};

InvestControl.prototype.handleIC04 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var sql = "SELECT ";
    var condition = "loan.name,invest.id,invest.create_time,invest.money,invest.loan_id,invest.status";
    var whCondition = " from invest LEFT JOIN loan  on invest.loan_id=loan.id where invest.status=1200 and (loan.status=1200 or loan.status=1300)  and invest.user_id='" + bodyNode.id + "'";
    var exSql = sql + condition + whCondition;
    exSql += '  order by invest.create_time desc';
    exSql += '  limit ' + bodyNode.skip;
    exSql += ',' + bodyNode.limit;
    conn.execute(exSql, [], function (err, data) {
        for(var key in data)
        {
            var set = data[key];
            set.create_time = dateUtil.toString(set.create_time);
        }
        var backNode = {};
        backNode.rst = data;
        var countSql=sql+" count(*)  as num "+whCondition;
        conn.execute(countSql, [], function (err, countData) {
            backNode.count = countData[0].num;
            cb(null,backNode)
        });
    });
};

/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  IC05 已结清投资
 */
InvestControl.prototype.checkIC05 = function (headNode, bodyNode, cb) {
    cb(null);
};

InvestControl.prototype.handleIC05 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var sql = "SELECT ";
    var condition = "loan.name,invest.id,invest.create_time,invest.money,(SELECT sum(invest_repay.interest) + sum(invest_repay.st_interest) FROM invest_repay WHERE invest_repay. STATUS = 1 AND invest.id = invest_repay.invest_id ) AS sumaCcrualMoney";
    var whCondition = " FROM invest LEFT JOIN loan ON invest.loan_id = loan.id WHERE invest. STATUS = 1200 AND loan. STATUS = 2000 AND invest.user_id ='" + bodyNode.id + "'";
    var exSql = sql + condition + whCondition;
    exSql += '  order by invest.create_time desc';
    exSql += '  limit ' + bodyNode.skip;
    exSql += ',' + bodyNode.limit;
    conn.execute(exSql, [], function (err, data) {
        for(var key in data)
        {
            var set = data[key];
            set.create_time = dateUtil.toString(set.create_time);
        }
        var backNode = {};
        backNode.rst = data;
        var countSql=sql+" count(*)  as num "+whCondition;
        conn.execute(countSql, [], function (err, countData) {
            backNode.count = countData[0].num;
            cb(null,backNode)
        });
    });
};

/**
 * @param headNode
 * @param bodyNode
 * @param cb
 * @Description  IC06 未成功投资
 */
InvestControl.prototype.checkIC06 = function (headNode, bodyNode, cb) {
    cb(null);
};

InvestControl.prototype.handleIC06 = function (headNode, bodyNode, cb) {
    var conn = dc.main.getConn();
    var sql = "SELECT ";
    var condition = "	loan.name,invest.create_time,invest.money,loan.deadline,invest.status";
    var whCondition = " FROM invest LEFT JOIN loan ON invest.loan_id = loan.id WHERE invest.status = 1300 AND invest.user_id ='" + bodyNode.id + "'";
    var exSql = sql + condition + whCondition;
    exSql += '  order by invest.create_time desc';
    exSql += '  limit ' + bodyNode.skip;
    exSql += ',' + bodyNode.limit;
    conn.execute(exSql, [], function (err, data) {
        for(var key in data)
        {
            var set = data[key];
            set.create_time = dateUtil.toString(set.create_time);
        }
        var backNode = {};
        backNode.rst = data;
        var countSql=sql+" count(*)  as num "+whCondition;
        conn.execute(countSql, [], function (err, countData) {
            backNode.count = countData[0].num;
            cb(null,backNode)
        });
    });
};


var investControl = new InvestControl();
module.exports = investControl;
