// mcp_engine/YcxhEngine.js
/**
 * Created by shiqm on 16-4-21.
 */
'use strict';

var async = require('async');
var service = require('mcp_service');
var calculateService = service.calculateService;

var cons = require('mcp_constants');
var calCons = cons.calculateCons;
var investCons=cons.investStatus;

var esut = require('easy_util');
var dateUtil = esut.dateUtil;

var util = require('mcp_util');
var dateMathUtil = util.dateMathUtil;

var dc = require('mcp_db').dc;


var YcxhEngine = function () {
    var self = this;
};

/**
 * point 0  放款后升息
 * point 1  即投即生息
 * @param loanObj
 * @param repayTypeObj
 * @param cb
 */
YcxhEngine.prototype.start = function (loanObj, repayTypeObj, cb) {
    var self = this;
    if (repayTypeObj.point == 0) {
        self.ptx(loanObj, repayTypeObj, false, cb);
    } else if (repayTypeObj.point == 1) {
        self.ptx(loanObj, repayTypeObj, true, cb);
    } else {
        cb('不支持的计息节点');
    }
};

/**
 * 先计算投资人利息，借款人利息为投资人利息之和
 * @param loanObj
 * @param repayTypeObj
 * @param jt
 * @param cb
 */

YcxhEngine.prototype.pt = function (loanObj, repayTypeObj, jt, cb) {
    var investTab = dc.main.get('invest');
    var cursor = investTab.find({loan_id: loanObj.id,status:investCons.bid_success}, {}, []).sort({create_time: 1});
    cursor.dateToString();
    cursor.count(function (err, count) {
        var startNum = 0;
        var i = 0;
        //总即投利息
        var allSt_interest = 0;
        //已计算的利息，为了给最后一个投资者计算利息，保证整体持平
        var hadInvest = 0;
        //还款日期
        var repay_day = dateMathUtil.dateAdd('d', loanObj.accrue_time, loanObj.deadline);
        console.log("投资数量:" + count);
        async.whilst(
            function () {
                return count > 0;
            },
            function (whileCb) {
                console.log("计算中...");
                cursor.limit(startNum, calCons.split_num);
                cursor.toArray(function (err, data) {
                    var invest_repayTab = dc.main.get('invest_repay');
                    async.eachSeries(data, function (investObj, callback) {
                        var interest = 0;
                        interest = calculateService.daysCal(loanObj.deadline, loanObj.rate, investObj.money);
                        //即投计算
                        var st_interest = 0;
                        if (jt) {
                            st_interest = calculateService.jt(investObj, loanObj.accrue_time);
                            allSt_interest += st_interest;
                        }
                        var cond = {};
                        cond.id = investObj.id + '_1';
                        cond.money = investObj.money;
                        cond.interest = interest;
                        cond.repay_day = repay_day;
                        cond.invest_id = investObj.id;
                        cond.st_interest = st_interest;
                        cond.create_time = dateUtil.getCurTime();
                        cond.period = 1;
                        cond.all_period = 1;
                        cond.status = 0; //0为还款  1已还
                        invest_repayTab.save(cond, [], function (err) {
                            investTab.update({id: investObj.id}, {
                                $set: {
                                    interest: interest,
                                    st_interest: st_interest
                                }
                            }, [], function (err) {
                                hadInvest += interest;
                                callback();
                            })
                        });
                    }, function (err) {
                        i++;
                        startNum = i * calCons.split_num;
                        count -= calCons.split_num;
                        whileCb();
                    });
                });
            },
            function (err) {
                if (!err) {
                    //todo 保存借款人还款记录
                    var loan_repayTab = dc.main.get('loan_repay');
                    var cond = {};
                    cond.id = loanObj.id + '_1';
                    cond.money = loanObj.money;
                    cond.interest = hadInvest;
                    cond.repay_day = repay_day;
                    cond.loan_id = loanObj.id;
                    cond.st_interest = allSt_interest;
                    cond.create_time = dateUtil.getCurTime();
                    cond.period = 1;
                    cond.all_period = 1;
                    cond.status = 0; //0为还款  1已还
                    loan_repayTab.save(cond, [], function (err) {
                        cb(true);
                    });
                } else {
                    cb(false);
                }
            }
        );
    });
};


/**
 * 先计算借款人利息，然后处理投资人利息，最后一名投资人利息为：借款人利息-其他投资人利息
 * @param loanObj
 * @param repayTypeObj
 * @param jt
 * @param cb
 */

YcxhEngine.prototype.ptx = function (loanObj, repayTypeObj, jt, cb) {
    var investTab = dc.main.get('invest');
    var cursor = investTab.find({loan_id: loanObj.id,status:investCons.bid_success}, {}, []).sort({create_time: 1});
    cursor.dateToString();
    cursor.count(function (err, count) {
        var startNum = 0;
        var i = 0;
        //正在处理的投资人顺序索引
        var order = 1;
        //总即投利息
        var allSt_interest = 0;
        //已计算的利息，为了给最后一个投资者计算利息，保证整体持平
        var hadInterest = 0;
        //总共的利息
        var sumInterest = calculateService.daysCal(loanObj.deadline, loanObj.rate, loanObj.money);
        //还款日期
        var repay_day = dateMathUtil.dateAdd('d', loanObj.accrue_time, loanObj.deadline);
        console.log("投资数量:" + count);
        async.whilst(
            function () {
                return count > 0;
            },
            function (whileCb) {
                console.log("计算中...");
                cursor.limit(startNum, calCons.split_num);
                cursor.toArray(function (err, data) {
                    var invest_repayTab = dc.main.get('invest_repay');
                    async.eachSeries(data, function (investObj, callback) {
                        var interest = 0;
                        //todo 最后一名
                        if (order == count) {
                            interest = sumInterest - hadInterest;
                        } else {
                            interest = calculateService.daysCal(loanObj.deadline, loanObj.rate, investObj.money);
                        }
                        //即投计算
                        var st_interest = 0;
                        if (jt) {
                            st_interest = calculateService.jt(investObj, loanObj.accrue_time);
                            allSt_interest += st_interest;
                        }
                        var cond = {};
                        cond.id = investObj.id + '_1';
                        cond.money = investObj.money;
                        cond.interest = interest;
                        cond.repay_day = repay_day;
                        cond.invest_id = investObj.id;
                        cond.st_interest = st_interest;
                        cond.create_time = dateUtil.getCurTime();
                        cond.period = 1;
                        cond.all_period = 1;
                        cond.status = 0; //0为还款  1已还
                        invest_repayTab.save(cond, [], function (err) {
                            investTab.update({id: investObj.id}, {
                                $set: {
                                    interest: interest,
                                    st_interest: st_interest
                                }
                            }, [], function (err) {
                                hadInterest += interest;
                                order++;
                                callback();
                            })
                        });
                    }, function (err) {
                        i++;
                        startNum = i * calCons.split_num;
                        count -= calCons.split_num;
                        whileCb();
                    });
                });
            },
            function (err) {
                if (!err) {
                    //todo 保存借款人还款记录
                    var loan_repayTab = dc.main.get('loan_repay');
                    var cond = {};
                    cond.id = loanObj.id + '_1';
                    cond.money = loanObj.money;
                    cond.interest = sumInterest;
                    cond.repay_day = repay_day;
                    cond.loan_id = loanObj.id;
                    cond.st_interest = allSt_interest;
                    cond.create_time = dateUtil.getCurTime();
                    cond.period = 1;
                    cond.all_period = 1;
                    cond.status = 0; //0为还款  1已还
                    loan_repayTab.save(cond, [], function (err) {
                        cb(true);
                    });
                } else {
                    cb(false);
                }
            }
        );
    });
};


var ycxhEngine = new YcxhEngine();
module.exports = ycxhEngine;
