// mcp_service/AdminService.js
/**
 * Created by shiqm on 16-3-24.
 */
'use strict';

var dc = require('mcp_db').dc;
var esul = require('easy_util');
var digestUtil = esul.digestUtil;
var util = require('mcp_util');
var objectUtil = util.objectUtil;
var async = require('async');

var AdminService = function () {
};

AdminService.prototype.login = function (username, password, cb) {
    var conn = dc.main.getConn();
    var tempPass = digestUtil.sha1(password);
    var tag = false;
    var sql = "SELECT * FROM user WHERE username='" + username + "'  and password='" + tempPass + "' and permission=1 and status=1";
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(err, null);
        } else {
            if (data.length == 1) {
                tag = true;
            }
            cb(err, tag);
        }
    })
};


AdminService.prototype.userLogin = function (username, password, cb) {
    var conn = dc.main.getConn();
    var tempPass = digestUtil.sha1(password);
    var tag = false;
    var sql = "SELECT * FROM user WHERE username='" + username + "'  and password='" + tempPass + "' and permission=0 and status=1";
    conn.execute(sql, [], function (err, data) {
        if (err) {
            cb(err, null);
        } else {
            if (data.length == 1) {
                tag = true;
            }
            cb(err, tag);
        }
    })
};


AdminService.prototype.getMenuByRoleId = function (roleId, cb) {
    var conn = dc.main.getConn();
    var sql = "SELECT menu_id from role_menu where role_id='" + roleId + "'";
    conn.execute(sql, [], function (err, data) {
        cb(err, data);
    });
};


AdminService.prototype.getMenu = function (username, cb) {
    var self = this;
    var conn = dc.main.getConn();
    if (!objectUtil.isNull(username)) {
        username = "and a.username='" + username + "' ";
    } else {
        username = '';
    }
    var sql =
        "SELECT DISTINCT  e.id," +
        "e.name," +
        "e.url," +
        "e.pid," +
        "e.status," +
        "e.type," +
        "e.level," +
        "e.seqnum " +
        "FROM user a " +
        "left join user_role b on a.id=b.user_id " +
        "left join role c on c.id=b.role_id " +
        "left join role_menu d  on c.id=d.role_id " +
        "left join menu e on d.menu_id=e.id " +
        "where e.type='management' " +
        "and e.status=1  " + username +
        "order by e.seqnum  asc";
    conn.execute(sql, [], function (err, allMenuArr) {
        if (err) {
            cb(err);
        } else {
            self.getMenuTree(allMenuArr, null, null, function (err, tempArr) {
                cb(err, tempArr);
            })
        }
    });
};

AdminService.prototype.getMenuTree = function (allMenuArr, menu, tempArr, cb) {
    var self = this;
    if (objectUtil.isNull(tempArr)) {
        tempArr = new Array();
    }
    var i = -1;
    async.eachSeries(allMenuArr, function (menuCategory, callback) {
        i++;
        if (objectUtil.isNull(menuCategory)) {
            callback(null);
        } else {
            var pid = menuCategory.pid;
            var menu_id = null;
            if (!objectUtil.isNull(menu)) {
                menu_id = menu.id;
            }
            if ((objectUtil.isNull(menu) && objectUtil.isNull(pid)) || (!objectUtil.isNull(menuCategory) && menu_id == pid )) {
                tempArr.push(menuCategory);
                delete allMenuArr[i];
                self.hasChildren(menuCategory, allMenuArr, function (err, tag) {
                    if (err) {
                        callback(err);
                    } else {
                        if (tag) {
                            self.getMenuTree(allMenuArr, menuCategory, tempArr, function (err, data) {
                                tempArr = data;
                                callback(null);
                            })
                        } else {
                            callback(null);
                        }
                    }
                });
            } else {
                callback(null);
            }
        }
    }, function (err) {
        cb(err, tempArr);
    });
};

AdminService.prototype.hasChildren = function (menu, allMenuArr, cb) {
    var tag = false;
    for (var key in allMenuArr) {
        var menuCategory = allMenuArr[key];
        if (menuCategory.pid == menu.id) {
            tag = true;
            break;
        }
    }
    cb(null, tag);
};


module.exports = new AdminService();
