// mcp_service/CreditorCalculateService.js
/**
 * Created by liujun on 2016/7/21.
 */
'use strict';
var cons = require('mcp_constants');
var calCons = cons.calculateCons;


var esut = require('easy_util');
var dateUtil = esut.dateUtil;

var util = require('mcp_util');
var dateMathUtil = util.dateMathUtil;
var creditorDateUtil = util.creditorDateUtil;

var CreditorCalculateService = function () {
};

/**
 * 等额本息
 * 债权价值
 */
CreditorCalculateService.prototype.creditorValueDeMonth = function (curRepayDay, curDay, preRepayDay, money, interest) {
    return creditorDateUtil.twoPrecise(money +((dateMathUtil.daysBetween(curDay, preRepayDay) - 1)/(dateMathUtil.daysBetween(curRepayDay, preRepayDay)
        )*interest));
};
/**
 * 一次性还本付息
 * 债权价值
 */
CreditorCalculateService.prototype.creditorValueYcxh = function (curRepayDay, curDay, ycxhAccurDate, money, interest) {
    return creditorDateUtil.twoPrecise(money + (dateMathUtil.daysBetween(curDay, ycxhAccurDate) - 1)/(dateMathUtil.daysBetween(curRepayDay, ycxhAccurDate)
        )*interest);
};

/**
 * 等额本息
 * 实际利率
 */
CreditorCalculateService.prototype.creditorActualRate = function (loanMoney, rate, repayMonth, period, tranMoney, actualMoney, surMoney) {
    return creditorDateUtil.twoPrecise(((loanMoney * (rate/12) - repayMonth) * (1 + Math.pow((rate/12),(period - 1))) + repayMonth)
        *tranMoney/surMoney/actualMoney*12);
};
/**
 * 一次性还本付息
 * 实际利率
 */
CreditorCalculateService.prototype.creditorYcxhActualRate = function (interest, ycxhDeadline, money, loanMoney, actualMoney) {
    return creditorDateUtil.twoPrecise(interest/ycxhDeadline*money/loanMoney/actualMoney*360)*100;
};

/**
 * 按月付息
 * 实际利率
 */
CreditorCalculateService.prototype.creditorAyfxActualRate = function (monthInteres, tranMoney, loanMoney, actualMoney) {
    return creditorDateUtil.twoPrecise(monthInteres * tranMoney/loanMoney * 12/actualMoney/100)*100;
}

/**
 * 等额本息
 * 受让人本金
 */
CreditorCalculateService.prototype.acceptmoney = function (money,total_price,moneyCur,surMoney) {
    return creditorDateUtil.twoPrecise((money/total_price)*((total_price/surMoney)*moneyCur));
}

/**
 * 等额本息
 * 受让人当期利息
 */
CreditorCalculateService.prototype.acceptinterestCur = function (curRepayDay,curDay,preRepayDay,money,total_price,interest,surMoney) {
    return creditorDateUtil.twoPrecise((money/total_price)*(total_price/surMoney*interest)*((dateMathUtil.daysBetween(curRepayDay, curDay) + 1)/(dateMathUtil.daysBetween(curRepayDay, preRepayDay))));
}

/**
 * 等额本息
 * 受让人利息
 */
CreditorCalculateService.prototype.acceptinterest = function (money,total_price,interest,surMoney) {
    return creditorDateUtil.twoPrecise((money/total_price)*(total_price/surMoney*interest));
}

var creditorCalculateService = new CreditorCalculateService();
module.exports = creditorCalculateService;
