// mcp_config/Prop.js
var esdb = require('easy_db');
var exports = {};
var target = 'dev';

var argv = process.argv;
var kvs = {};
for(var key in argv)
{
    if(key > 1)
    {
        var kv = argv[key].split("=");
        kvs[kv[0]] = kv[1];
    }
}
if(kvs.target)
{
    target = kvs.target;
}

//网关端口
if(kvs.gtPort)
{
    exports.gtPort = kvs.gtPort;
}
else
{
    exports.gtPort = 9090;
}

//队列名词
if(kvs.mq)
{
    exports.queue = kvs.mq;
}
else
{
    exports.queue = 'backWay'+kvs.gtPort;
}

exports.platform = {};

//runtime target
exports.target = target;

//config db basic type
var dbs = {
    'testMg': {
        config: {'url': 'mongodb://192.168.1.201:27017/node_gsyd'},
        type: esdb.prop.dbType.mongodb
    },
    'testMain': {
        config: {
            'host': '192.168.1.201',
            'user': 'root',
            'password': '123456',
            'port': 3306,
            'database': 'node_mcp'
        },
        type: esdb.prop.dbType.mysql,
        dateToLong: false,
        poolSize:4
    },
    'testMq': {
        host: '127.0.0.1',
        port: 5672
    },

    'devMg': {
        config: {'url': 'mongodb://10.144.80.133:27017/node_gsyd'},
        type: esdb.prop.dbType.mongodb
    },
    'devMain': {
        config: {
            'host': '127.0.0.1',
            'user': 'root',
            'password': 'boyarenhe888',
            'port': 3306,
            'database': 'node_mcp'
        },
        type: esdb.prop.dbType.mysql,
        dateToLong: false,
        poolSize:4
    },
    'devMq': {
        host: '10.144.80.133',
        port: 5672
    },

    'runMg': {
        config: {'url': 'mongodb://10.25.108.247:27017/node_gsyd'},
        type: esdb.prop.dbType.mongodb
    },
    'runMain': {
        config: {
            'host': '10.45.231.114',
            'user': 'node',
            'password': 'gB4gu7xAxd',
            'port': 3306,
            'database': 'node_mcp'
        },
        type: esdb.prop.dbType.mysql,
        dateToLong: false,
        poolSize:20
    },
    'runMq': {
        host: '127.0.0.1',
        port: 5672
    }
};


//config http basic type
if(target == 'dev')
{
    exports.main = dbs.devMain;
    exports.mg = dbs.devMg;
    exports.mq = dbs.devMq;
    exports.platform.site = {
        hostname: '127.0.0.1',
        port: 9088,
        path: '/gs-filter/main/interface.htm',
        method: 'POST'
    };

    exports.platform.gateways = [
        {host:'127.0.0.1', port:9090, method:'POST'},
        {host:'127.0.0.1', port:9091, method:'POST'}
    ];

    exports.filterPort = 9088;
}
else if(target == 'test')
{
    exports.main = dbs.testMain;
    exports.mg = dbs.testMg;
    exports.mq = dbs.testMq;
    exports.platform.site = {
        hostname: '127.0.0.1',
        port: 9088,
        path: '/gs-filter/main/interface.htm',
        method: 'POST'
    };
    exports.platform.gateways = [
        {host:'127.0.0.1', port:9090, method:'POST'},
        {host:'127.0.0.1', port:9091, method:'POST'}
    ];
    exports.filterPort = 9088;
}
else if(target == 'run')
{
    exports.main = dbs.runMain;
    exports.mg = dbs.runMg;
    exports.mq = dbs.runMq;
    exports.platform.site = {
        hostname: '127.0.0.1',
        port: 9088,
        path: '/gs-filter/main/interface.htm',
        method: 'POST'
    };
    exports.platform.gateways = [
        {host:'127.0.0.1', port:9090, method:'POST'},
        {host:'127.0.0.1', port:9091, method:'POST'},
        {host:'127.0.0.1', port:9092, method:'POST'},
        {host:'127.0.0.1', port:9093, method:'POST'}
    ];
    exports.filterPort = 9088;
}

if(kvs.filterPort)
{
    exports.filterPort = kvs.filterPort;
}


//暂时支持3种密钥来源
exports.digestFromType = {"NONE":0, "DB":1, "CACHE":2, "FIX":3};
exports.digestFromTypeArray = [
    {id:0, code:'NONE', des:"无"},
    {id:1, code:'DB', des:"数据库"},
    {id:2, code:'CACHE', des:"缓存"},
    {id:3, code:'FIX', des:"缓存固定"}
];

//if user hasn't operation in half a hour, the key will be expired.

module.exports = exports;
