// mcp_constants/UserBillCons.js
var UserBillCons = function () {
    var self = this;
    self.transType = {
        FROZEN: "frozen",   // description:"冻结"
        THAW: "thaw",  // description:"解冻"
        OUT: "out",   // description:"从余额转出"
        IN: "in",  // description:"转入到余额"
        FROZEN_OUT: "frozen_out"  // description:"从冻结金额中转出"
    };

    self.payType = {
        INVEST: "invest",   // description:"投资成功"
        ADMIN_OPERATION: "admin_operation",  // description:"管理员干预"
        RECHARGE_SUCCESS: "recharge_success",   // description:"充值成功"
        APPLY_LOAN: "apply_loan",  // description:"申请借款"
        REFUSE_APPLY_LOAN: "refuse_apply_loan",  // description:"借款申请未通过"
        APPLY_WITHDRAW: "apply_withdraw",  // description:"申请提现"
        REFUSE_APPLY_WITHDRAW: "refuse_apply_withdraw",  // description:"提现申请未通过"
        NORMAL_REPAY: "normal_repay",  // description:"正常还款"
        ADVANCE_REPAY: "advance_repay",  // description:"提前还款"
        OVERDUE_REPAY: "overdue_repay",  // description:"逾期还款"
        CANCEL_LOAN: "cancel_loan",  // description:"借款流标"
        WITHDRAW_LOAN: "withdraw_loan",  // description:"借款撤标"
        GIVE_MONEY_TO_BORROWER: "give_money_to_borrower",  // description:"借款放款"
        WITHDRAW_SUCCESS: "withdraw_success",  // description:"提现成功"
        CANCEL_INVEST: "cancel_invest",  // description:"投资流标"
        TRANSFER: "transfer",  // description:"债权转让成功"
        TRANSFER_BUY: "transfer_buy",  // description:"债权购买成功"
        USE_COUPON: "use_coupon",  // description:"红包使用"
        FEE: "fee"  // description:"红包使用"
    };



};

module.exports = new UserBillCons();
